TagsCloudVisualization/TagCloudTests/CircularCloudLayouterTests.cs

```diff
@@ -0,0 +1,174 @@
+    {
+        var cases = new List<TestCaseData>();
+        for (int x = -1; x <= 1; x++)
+        {
+            for (int y = -1; y <= 1; y++)
+            {
+                var name = "HasOverlapWith: " + (x < 0 ? "Left " : (x > 0) ? "Right " : "Center ")
+                         + (y < 0 ? "Top " : (y > 0) ? "Bottom " : "Center ");
+                const int sideLength = 6;
+                cases.Add(new TestCaseData(sideLength * x - 2, sideLength * y - 2, sideLength, sideLength).Returns(true).SetName(name + "overlap"));
+
+                if(x != 0 && y != 0)
+                    cases.Add(new TestCaseData((sideLength + 20) * x, (sideLength + 20) * y, sideLength, sideLength).Returns(false).SetName(name + "not overlap"));
+            }
+        }
+
+        cases.Add(new TestCaseData(-10, -10, 20, 40).Returns(true).SetName("HasOverlapWith: Area inside targer area"));
+        cases.Add(new TestCaseData(-6, -6, 12, 12).Returns(true).SetName("HasOverlapWith: The same area"));
+        return cases;
+    }
+
+    [TestCaseSource(nameof(GetOverlapTests))]
+    public bool HasOverlapWith_CorrectAnswer_OnOverlap(int xOffset, int yOffset, int width, int height)
+    {
+        var staticSize = new Size(12, 12);
+        layouter.PutNextRectangle(staticSize);
+        var rectangle = new Rectangle(new Point(center.X + xOffset, center.Y + yOffset), new Size(width, height));
+        return layouter.HasOverlapWith(rectangle);
+    }
+
+    [TestCase(10)]
+    [TestCase(100)]
+    [Repeat(50)]
+    public void PutNextRectangle_Error_OnOverlap(int wordCount)
+    {
+        for (int i = 0; i < wordCount; i++)
+            layouter.PutNextRectangle(new Size(50 + random.Next(-20, 70), 30 + random.Next(-20, 20)));
+
+        var rectengles = layouter.Rectangles;
+        for (int i = 0; i < rectengles.Count - 1; i++)
+        {
+            var rect = rectengles[i];
+            for (int j = i + 1; j < rectengles.Count; j++)
+            {
+                var otherRect = rectengles[j];
+                if (otherRect.IntersectsWith(rect))
+                    throw new Exception("Rectangles overlap!");
+            } 
+        }
+    }
+
+    [Test]
+    public void PutNextRectangle_IsСloudDensely()
+    {
+        const double accuracy = 0.2;
+        layouter.Density = 0.01;
+
+        // The more rectangles, the more accurately repeats the shape of a circle
+        for (int i = 0; i < 200; i++)
+            layouter.PutNextRectangle(new Size(50, 20));
+
+        var rects = layouter.Rectangles;
+        var rectangleArea = rects.Select(x => x.Width * x.Height).Sum();
+        var horizontalRadius = Math.Abs(rects.Max(x => x.Right) - rects.Min(x => x.Left)) / 2 ;
+        var verticalRadius = Math.Abs(rects.Max(x => x.Bottom) - rects.Min(x => x.Top)) / 2;
+        var boundingElipseArea = Math.PI * horizontalRadius * verticalRadius;
+        (Math.Abs(rectangleArea / boundingElipseArea - 1) < accuracy).Should().BeTrue();
```

Comment:
В FluentAssertions есть методы для сравнения чисел:
```suggestion
(rectangleArea / boundingElipseArea - 1).Should().BeLessThan(accuracy);
```